cdap-data-fabric/src/main/java/co/cask/cdap/data/format/CombinedLogRecordFormat.java

```diff
@@ -0,0 +1,149 @@
+import co.cask.cdap.api.flow.flowlet.StreamEvent;
+import com.google.common.base.Charsets;
+import com.google.common.collect.Lists;
+
+import java.nio.charset.Charset;
+import java.util.Iterator;
+import java.util.List;
+
+/**
+ * Stream record format that interprets stream body as data in Combined Log Format.
+ * CLF format: remote_host remote_login auth_user [date] "request" status content_length referrer user_agent
+ * Sample CLF data:
+ * 220.181.108.77 - - [01/Feb/2015:06:59:57 +0000] "GET / HTTP/1.1" 301 295 "-"
+ * "Mozilla/5.0 (compatible; Baiduspider/2.0; +http://www.baidu.com/search/spider.html)"
+ *
+ */
+public class CombinedLogRecordFormat extends StreamEventRecordFormat<StructuredRecord> {
+  private Charset charset = Charsets.UTF_8;
+
+  @Override
+  public StructuredRecord read(StreamEvent event) throws UnexpectedFormatException {
+    String bodyAsStr = Bytes.toString(event.getBody(), charset);
+    StructuredRecord.Builder builder = StructuredRecord.builder(schema);
+    List<String> parts = getLogEntries(bodyAsStr);
+    List<Schema.Field> fields = schema.getFields();
+    int index = 0;
+    while (index < fields.size()) {
+      Schema.Field field = fields.get(index);
+      String val = (parts.size() < index || (parts.get(index).equals("-") &&
+                                             field.getSchema().getType() != Schema.Type.STRING))
+                    ? null : parts.get(index);
+      builder.convertAndSet(fields.get(index).getName(), val);
+      index++;
+    }
+
+    return builder.build();
+  }
+
+
+  @Override
+  protected Schema getDefaultSchema() {
+    return Schema.recordOf("streamEvent",
+           Schema.Field.of("remote_host", Schema.unionOf(Schema.of(Schema.Type.STRING), Schema.of(Schema.Type.NULL))),
+           Schema.Field.of("remote_login ", Schema.unionOf(Schema.of(Schema.Type.STRING), Schema.of(Schema.Type.NULL))),
+           Schema.Field.of("auth_user", Schema.unionOf(Schema.of(Schema.Type.STRING), Schema.of(Schema.Type.NULL))),
+           Schema.Field.of("date", Schema.unionOf(Schema.of(Schema.Type.STRING), Schema.of(Schema.Type.NULL))),
+           Schema.Field.of("request", Schema.unionOf(Schema.of(Schema.Type.STRING), Schema.of(Schema.Type.NULL))),
+           Schema.Field.of("status", Schema.unionOf(Schema.of(Schema.Type.INT), Schema.of(Schema.Type.NULL))),
+           Schema.Field.of("content_length", Schema.unionOf(Schema.of(Schema.Type.INT), Schema.of(Schema.Type.NULL))),
+           Schema.Field.of("referrer", Schema.unionOf(Schema.of(Schema.Type.STRING), Schema.of(Schema.Type.NULL))),
+           Schema.Field.of("user_agent", Schema.unionOf(Schema.of(Schema.Type.STRING), Schema.of(Schema.Type.NULL))));
+  }
+
+
+  @Override
+  protected void validateSchema(Schema desiredSchema) throws UnsupportedTypeException {
+    // a valid schema is a record of simple types.
+    Iterator<Schema.Field> fields = desiredSchema.getFields().iterator();
```

Comment:
We can use clf to define common log format as well. in which case we will pass in the schema explicitly. Common log format is a subset of combined log format without the referrer and useragent.